HeavyFlavor Validation update

tmadlener opened the pull request:
Necessary updates to have meaningful/filled validation plots again for the HeavyFlavor validation package (used by BPH for RelVal):

- Update the trigger paths to the 2017 HLT menu. Where possible the paths from 2016 have been updated to their 2017 version. Some paths don't have a 2017 version. These have been removed.
- Refactored the assignment of filter levels to the different filters (based on the filter name) to make it more stable and to get rid of newly introduced duplicates due to the IterL3 modules in the paths.

tmadlener committed:
Add 2017 paths and remove obsolete 2016 paths.

tmadlener committed:
Use baseFolderPath variable also for remaining 2016 modules.

tmadlener committed:
Make filter level assignment based on module names a bit more stable.

Internally used filter level assignment relied on the name of the modules,
which can be ambiguous. Added an additional check on the type of the module
to be a bit more stable and to prevent from multiple modules on the same level.

The newly introduced function yields the same levels as the previous definition
in dqmBeginRun, without having multiple modules on the same level.

Removed some trailing whitespace as well.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20520/705

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/PR-20520/705/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/PR-20520/705/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly (this will soon be required for PRs to be considered)

tmadlener committed:
Apply code-checks patch

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20520/749

cmsbuild commented:
A new Pull Request was created by  @tmadlener (Thomas Madlener) for master.

It involves the following packages:

HLTriggerOffline/HeavyFlavor


@kmaeshima, @cmsbuild, @vanbesien, @vazzolini, @dmitrijus can you please review it and eventually sign? Thanks.
@mtosi this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mtosi commented:
@cmsbuil, please test

mtosi commented:
@tmadlener first of all, thanks !
then, may I suggest to update the directory into BPH ?

tmadlener commented:
@mtosi which directory are you referring to? The output directory of the plots or the package directory inside cmssw?

The first one is no problem. Only one parameter in the config file has to be changed for that and according to my quick check there is currently no conflicting BPH folder in the output directory.

The second one is a bit more work, but should be possible as well. However, I would have to check how many files would have to be changed and some of them probably would also need to be renamed. You can probably judge better than me how much additional work would be involved in that (just the renaming, etc. part).

mtosi commented:
sorry the confusion
I mean

baseFolderPath = 'HLT/HeavyFlavor/HLT/'
 -->

baseFolderPath = 'HLT/BPH/'

thanks
 mia

On Mon, Sep 18, 2017 at 2:01 PM, Thomas Madlener <notifications@github.com>
wrote:

> @mtosi <https://github.com/mtosi> which directory are you referring to?
> The output directory of the plots or the package directory inside cmssw?
>
> The first one is no problem. Only one parameter in the config file has to
> be changed for that and according to my quick check there is currently no
> conflicting BPH folder in the output directory.
>
> The second one is a bit more work, but should be possible as well.
> However, I would have to check how many files would have to be changed and
> some of them probably would also need to be renamed. You can probably judge
> better than me how much additional work would be involved in that (just the
> renaming, etc. part).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/20520#issuecomment-330199084>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEt581gbJaDWDHmZqJYGTQTupcjOr1Nhks5sjluTgaJpZM4PYHWP>
> .
>

tmadlener committed:
Change output directory to BPH.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20520/766

cmsbuild commented:
Pull request #20520 was updated. @kmaeshima, @cmsbuild, @vanbesien, @vazzolini, @dmitrijus can you please check and sign again.

dmitrijus commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

tmadlener commented:
Hi all,
I have a question about the backports. Will this PR be automatically backported to 9_2_X or is this something we have to request (or do ourselves)?

Cheers, Thomas

mtosi commented:
there are not "automatic backport"
  mia

On Tue, Sep 19, 2017 at 4:32 PM, Thomas Madlener <notifications@github.com>
wrote:

> Hi all,
> I have a question about the backports. Will this PR be automatically
> backported to 9_2_X or is this something we have to request (or do
> ourselves)?
>
> Cheers, Thomas
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/20520#issuecomment-330558927>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEt582nwBGsTxCz555RPceM2O3AuwxlOks5sj9B_gaJpZM4PYHWP>
> .
>

tmadlener commented:
I see. So to clarify. What are the actions that have to be taken from our side to have a backport of this PR?

dmitrijus commented:
You have to make a pull request to the target branch.

And you should provide a good reason to have it backported - the PR will be touching production code.

tmadlener commented:
Thanks for the prompt answers.
Another question: Will this PR be in the next 9_3_0 pre-release (if there is one) or is a backport to 9_3_X necessary as well for that?

mtosi commented:
for the TSG is not mandatory
(but I would suggest to do so)

On Tue, Sep 19, 2017 at 5:01 PM, Thomas Madlener <notifications@github.com>
wrote:

> Thanks for the prompt answers.
> Another question: Will this PR be in the next 9_3_0 pre-release (if there
> is one) or is a backport to 9_3_X necessary as well for that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/20520#issuecomment-330568325>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEt58ylDhhP22d5OSffDAcbJ0K6IV3taks5sj9dcgaJpZM4PYHWP>
> .
>

tmadlener commented:
Ok, thanks for your patience and answers. :)
Since there are probably more validation campaigns to come for 9_2_X and 9_3_X releases, I will open PRs targetting these two branches as soon as this one has been merged, if there are no objections.

Having these changes also in those branches allows us to use the DQM GUI again for upcoming validation campaigns. Currently we have to run the validation and harvesting modules manually for each release, which gets tedious.

cmsbuild commented:
-1

Tested at: bd0826f1d229bd26f89ebad965a4a123eca98029


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e133687c2dd23771612485bf4baf046b9317e8e8
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Entering Package HLTriggerOffline/HeavyFlavor
Entering library rule at HLTriggerOffline/HeavyFlavor
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorHarvesting.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/PlotCombiner.cc 
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc:685:18: error: constexpr variable 'l3Filters' must be initialized by a constant expression
  constexpr auto l3Filters = {"HLTMuonDimuonL3Filter", "HLTMuonL3PreFilter"};
                 ^           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc:685:18: note: pointer to subobject of temporary is not a constant expression
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc:685:30: note: temporary created here
  constexpr auto l3Filters = {"HLTMuonDimuonL3Filter", "HLTMuonL3PreFilter"};
</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e133687c2dd23771612485bf4baf046b9317e8e8
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/git-merge-result

tmadlener committed:
Fix clang errors by replacing constexpr char[] by const std::string.

clang complained about arrays of chars not being initialized by a
constant expression. (gcc compiles without warning on this)

Subsequently the call to 'isAnyOf' does not succeed due to:
no known conversion from 'const initializer_list<const char *>' to 'const initializer_list<const const char *>' for 2nd argument

Removing the second const from const char* const would allow clang to compile
but breaks the gcc build due to the same reasoning (but with reversed arguments in this case).

Since this is called only in setup I have simply replaced all constexpr char[] by std::string.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20520/829

cmsbuild commented:
Pull request #20520 was updated. @kmaeshima, @cmsbuild, @vanbesien, @vazzolini, @dmitrijus can you please check and sign again.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2625349
 - DQMHistoTests: Total failures: 244
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2624916
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 14 edm output root files, 26 DQM output files

dmitrijus commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
+1
Tested at: 09217a6ef8a3a490e647543cf1f7048a310aef57
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23115/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/5e53c2b9539aa99597276fc2af455b7acffca3cf
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23115/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23115/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23115/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2625349
 - DQMHistoTests: Total failures: 235
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2624925
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 14 edm output root files, 26 DQM output files

davidlange6 commented on HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc:
```diff
@@ -145,43 +158,29 @@ void HeavyFlavorValidation::dqmBeginRun(const edm::Run& iRun, const edm::EventSe
 		return;
 	}else{
+          	stringstream os;
+                for (const auto &filters : filterNamesLevels) os << " " << filters.first;
```
hi @tmadlener - you can avoid the string stream here - just str=str+" "+filters.first in the loop

davidlange6 commented on HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc:
```diff
@@ -649,6 +648,71 @@ void HeavyFlavorValidation::myBook1D(DQMStore::IBooker & ibooker, TString name,
+        hltConfig.moduleType(moduleName) == l1Filter) {
+      level = 1;
+    }
+    if (contains(moduleName, "L2") && hltConfig.moduleType(moduleName) == l2Filter) {
```
can this and subsequent if statements be else if instead? eg, once you have found a level value, the subsequent ifs do not need to be checked, right?

tmadlener committed:
Implement Davids comments from code review.

- Avoid usage of stringstream where concatenating strings is possible.
- Make filter level assignment stable enough to allow for early returns.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20520/947

cmsbuild commented:
Pull request #20520 was updated. @kmaeshima, @cmsbuild, @vanbesien, @vazzolini, @dmitrijus can you please check and sign again.

davidlange6 commented:
please test


On 25 Sep 2017, at 09:49, cmsbuild <notifications@github.com<mailto:notifications@github.com>> wrote:


Pull request #20520<https://github.com/cms-sw/cmssw/pull/20520> was updated. @kmaeshima<https://github.com/kmaeshima>, @cmsbuild<https://github.com/cmsbuild>, @vanbesien<https://github.com/vanbesien>, @vazzolini<https://github.com/vazzolini>, @dmitrijus<https://github.com/dmitrijus> can you please check and sign again.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/20520#issuecomment-331870394>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEzyw6anZd2yBQGNlBZFMpsuizpznyRWks5sl6FKgaJpZM4PYHWP>.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: c098acf717046ef3952916711463b6b43af44fa5
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23208/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/630acaf4d413c283cf6a9ae3be6037f071b994cb
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23208/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23208/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23208/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2667964
 - DQMHistoTests: Total failures: 230
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2667545
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 15 edm output root files, 26 DQM output files

davidlange6 commented:
merge
